technologies/lambda/api-gateway-errors.md

```diff
@@ -0,0 +1,127 @@
+# Custom error codes with Lambda and API Gateway
+
+Combining Amazon's Lambda and API Gateway is an effective means of creating a robust and cost effective api.
+
+Lambda functions contains two primary methods called for success or failure but one quickly discovers  that ```context.fail()``` still returns a successful 200 response.  Ultimately both code paths below return 200.
+
+```js
+exports.myHandler = function(event, context) {
+   if (superAwesomeCode() === 'ok') {
+     context.succeed('It worked!');
+   } else {
+     context.fail('There was an error')
+   }
+}
+```
+
+This is obviously not a very helpful response for our api consumers. Fortunately Amazon has provided a solution for this (if not a slightly unusual one - but one that does lets us achieve what we need).
+
+### What do lambda error responses look like via the api gateway?
+
+Lets have a look at how failure calls from Lambda are displayed. The simple lambda code below...
+```js
+exports.myHandler = function(event, context) {
+  var error = new Error("Invalid data")
+  context.fail(error)
+}
+```
+...will result with the following response _(including the http status for clarity)_
+```js
+Status: 200
+{
+  "errorMessage": "Invalid data",
+  "errorType": "Error",
+  "stackTrace": [
+    "exports.handler (/var/task/index.js:11:17)"
```

Comment:
... so I looked at the standard node setup boilerplate projects in aws, they all have the pattern of exporting a method called "handler" via exports.handler.  So thinking of using that as the default? (but still removing all the "myHandler" with "handler".  Thoughts?